### ISSUE TITLE
Number of Iteration for simulateResiduals-function 

### ISSUE THREAD

**Issue by Hallo951** [florianhartig/DHARMa#76]
Hi,

I tested the simulateResiduals-function with various n and yielded strongly varying results. Even a slight variation of n (e.g. n=250  to n=254) showed a large range in KS test of the QQ plot residuals
(between 0.36 and 0.83) and I found no obvious regularities. Same goes for the Residual vs. predicted values (ranked transformed) plot. Why is that and which number of n should I choose? I would expect that the results become more robust with greater n without any erratic changes in between.

```
testData <-  createData(sampleSize = 100, fixedEffects = 2, family = binomial(), randomEffectVariance = 0)
m <- glmmTMB(observedResponse ~ Environment1 , data=testData, `family=binomial)
simulateResiduals(m,n = 250, plot = T)
simulateResiduals(m,n = 251, plot = T)
simulateResiduals(m,n = 252, plot = T)
simulateResiduals(m,n = 253, plot = T)
simulateResiduals(m,n = 254, plot = T)
simulateResiduals(m,n = 1000, plot = T)
simulateResiduals(m,n = 1001, plot = T)
```
Best Frank

**Comment by florianhartig** [comment]
This is expected behavior, but possibly a bit counterintuitive to the user. 

When calculating the simulated residuals, there are two sources of stochasticity in the simulations

- Simulating new data
- Smoothening out integer values 

Together they can create a substantial amount of stochasticity in simulated residuals. Most of this will disappear for large n, but there is a remaining part that is particular large for binomial models, where you have only 0 / 1 observed, and 0 / 1 simulated. To get the residuals homogenous between 0,1, they are basically smoothly distributed across the 0/1 space (this is the Dunn-Smyth idea). For the simulations, this averages out for high n, but there is a part for the observed data that won't be averaged out (because, say, your model expects a 0.5, and you observe a 1 - there is no way of assigning this residual a fixed, unique value, because if you'd do this for any observation with the same combination, you would get a line), so this part of the residuals remains stochastic regardless of n. 

Bottomline: this type of simulated residuals are inherently stochastic. 

Now, as stated in the vignette, section "Reproducibility notes, random seed and random state", by default, DHARMa fixes the random seed, so that running the same command 2 times will produce the same output. I did this to avoid that users re-run the residuals / tests, and pick the one that suits them best. In the way it is now, one possible combinations of random residuals will be calculated, and you are supposed to stick with it (at least if you want to ensure correct type I error rates of the tests).

If you change n, however, more random seeds are drawn for the simulations, and then you everything will change. This is what's happening here. 

If you want to see the same with fixed n, you can run 

`simulateResiduals(fittedModel, seed = F)`

**Comment by Hallo951** [comment]
Hi,

Thank you for your answer but I do not understand that completely.

I use the DHARMa package to check different glmmmTMB models (e.g. 1000 models) which all based on the same dataset, family and all are normaly (not zero-inflated). In other words, with glmmTMB, I create different variations (additive and multiplicative) of environmental variables to explain the same species.

I check these models with DHARMa. My goal is to identify all models that are ok after the simulateResiduals function. My problem is when i check e.g. 10 different models (which created after the above described pattern) with n = 250 three models are ok. By n = 252, five different models are ok and by n = 254 one model is ok. 

I always get different results depending on n and can not make a concrete statement about which of my models are really ok and which are not.

How can I solve this problem or what do I do best?

**Comment by florianhartig** [comment]
OK, let me rephrase it like that - the n really doesn't matter that much - in doubt, just to be sure, set n = 1000. 

What matters is that the tests in DHARMa are stochastic - if you have a **perfect model**, the residuals will still vary stochastically, so think of this as DHARMa running a new experiment on your model to find out if the model fits or not. 

Currently, I "switched" the stochasticity off, so that users don't play around with different repetitions, but if you want to switch it back on, you can do repeated

simulateResiduals(fittedModel, seed = F)

To see how much the residuals change. From my experience, I would suggest that if you see a clear pattern, then this is usually stable when rerunning the residuals. Of course, smaller things change, but these are then usually due to stochasticity. I had started to experiment with implementations of the residuals where this is done automatically, but there are a bunch of problems with then coming up with properly defined residuals / p-values, so I dropped this idea. But on a qualitative level, you can have a look if the things that you identify as a problem are stable when re-simulating. 

Now, this can of course not be done for 1000 models. But to be honest, I'm also a bit skeptical about the sense of doing this - residual checks are not meant as a substitute for model selection - if you want to select the best model, why not just use AIC (with all it's problems, in particular for inference, but I don't think residual checks are better for selecting adequate models).

The idea of residual checks is to do a lot of them (qq, res ~ predictors, res ~ predictors, REs, group) to check that your model is acceptable. For a model with a lot of data, even having significant qq plots may not make the model unacceptable. For an lm with > 1000 observations, it's uncommon to have perfect QQ plots. The residual checks are meant as a qualitative tool to explore potential problems, and it's then up to the analyst to decide if something needs to be done about this.

**Comment by Hallo951** [comment]
Aha ... thanks for the good explanation.

As for my idea, I wanted to check all 1000 models with DHARM and select those models that are ok. From the selected models I wanted to take the best model by AIC. Quasi the motto: "Check which of the models is internally ok and then take the best model after the AIC".

Did I understand correctly that they would turn this process around? To filter all 1000 models via AIC and then check the best models with DHARMa? If this is your idea, I have tested it. The result was not good. Most of the best models via AIC was in the QQ plot good, but in the Residual Vs. Predicted lines - Plot bad (quantile regression lines were in most cases parabolic and had no similarity to the quantile linines). In most cases, all the best models by AIC after DHARMa were not ok (QQ plot: ok, second plot: not ok -> in sum: not ok).

One more question about the Residual Vs. Predicted lines - Plot. Is there a measure that expresses the plot with the quantile regression is ok? I can not visually check the plot of 1000 models. Currently I've got myself with the following code:

```
sresid <- simulateResiduals(mod1,n=1000)
plot(sresid)

res = sresid$scaledResiduals
pred = sresid$fittedPredictedResponse

# rank tranforms the predictor (if rank == T -> default)
pred = rank(pred, ties.method = "average")
pred = pred / max(pred) 
 
# Korrelation between res and pred
cor_test <- cor.test(res,pred, method = "kendall")
cor_test$p.value
```

**Comment by florianhartig** [comment]
Sorry, looking through the open issues, I realize I had forgotten to answer this question. Afew short comments, for completeness

I find it somewhat surprising that AIC and residual checks go in such different directions. Are you sure this is really systematic?

One problem I have noticed from questions (e.g. https://github.com/florianhartig/DHARMa/issues/77) is that people read too much into the quantile regression lines. If you have only 10 data points, these lines can be all over the place. Maybe I should just change to a spline and let people judge the plot visually. Would you say you see a pattern visually?

As for formal tests - if you have the option rank = T (rank transformation on x) a perfect fit should be random in [0,1]x[0,1] space, i.e. should follow a poisson point process. I have played around with the idea of measuring the deviation from a poisson process, which would give you a value and also a p-value, but I haven't implemented this yet. 

Will close this for now!
